### PR TITLE
SpecDec Bench: February Update

### PR DESCRIPTION
**Pull request by IzzyPutterman**
## What does this PR do?

**Type of change:** ? 

**Overview:** 
Addition of SpecBench Dataset
Addition of NVIDID SPEED-Bench dataset, preproc scripts, and custom metrics aggregator
Addition of example of converting SpecBench Medusa to this FW
Addition of Initial TRTLLM AutoDeploy Specdec support

Updates to all frameworks for better perf (overlap/async scheduling etc)

## Usage

```python
# Add a code snippet demonstrating how to use this
```

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes/No 
- **Did you write any new necessary tests?**: Yes/No
- **Did you add or update any necessary documentation?**: Yes/No
- **Did you update [Changelog](https://github.com/NVIDIA/Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes/No 

## Release Notes

* **New Features**
  * Added SPEED-Bench dataset support with configurable throughput and qualitative configurations
  * Introduced SpecBench metrics with acceptance rate analysis and visualizations
  * Added progress bar during benchmark execution
  * New model implementations for auto-deployment and Medusa-style speculative decoding
  * Data preparation utility for benchmark datasets
  * Enhanced metrics with per-category analysis and performance charts

* **Documentation**
  * Updated README with SPEED-Bench workflow and examples
  * New porting guide for integrating custom benchmark runners

* **Refactor**
  * Streamlined model and runner interfaces for improved flexibility
  * Consolidated dataset implementations and removed deprecated base classes

* **Chores**
  * Added required dependencies for data handling and visualizations

### CODE REVIEW

**Comment by yeyu-nvidia** [comment]
Can you ask some description on what this PR is for?

**Comment by IzzyPutterman** [comment]
> Can you ask some description on what this PR is for?

Added info to the top!

File: examples/specdec_bench/README.md [hunk 1]
```diff
@@ -41,6 +41,52 @@ python3 run.py --model_dir openai/gpt-oss-120b --tokenizer openai/gpt-oss-120b -
...
+ADDITIONAL INFORMATION: MIT for bigcode/humanevalpack, RUCAIBox/MMATH, RUCAIBox/BAMBOO and EQ-Bench. Apache 2.0 for Writing Bench and Spec-Bench. CC BY 4.0 for FBK-MT/MCIF. MIT and Apache 2.0 for tianyang/repobench_python_v1.1, JetBrains-Research/lca-project-level-code-completion and tianyang/repobench_java_v1.1.
+
+NOTICE: For each dataset a user elects to use, the user is responsible for checking if the dataset license is fit for the intended purpose. The `prepare_data.py` script automatically fetches data from all the source datasets.
+
+Additional details are in [HuggingFace dataset repository](https://huggingface.co/datasets/nvidia/SPEED-Bench).
+
+#### Qualitative split
+
+```bash
+python3 run.py --model_dir meta-llama/Llama-3.3-70B-Instruct --tokenizer meta-llama/Llama-3.3-70B-Instruct --draft_model_dir yuhuili/EAGLE3-LLaMA3.3-Instruct-70B --dataset speed --dataset_path data/speed/qualitative --tp_size 8 --ep_size 1 --draft_length 3 --output_length 4096 --engine TRTLLM --concurrency 32 --show_progress
```
Comment (benchislett) [inline examples/specdec_bench/README.md]: I don't think the examples in this README actually need to provide a path since they load from HF, right?
```suggestion
python3 run.py --model_dir meta-llama/Llama-3.3-70B-Instruct --tokenizer meta-llama/Llama-3.3-70B-Instruct --draft_model_dir yuhuili/EAGLE3-LLaMA3.3-Instruct-70B --dataset speed --dataset_path qualitative --tp_size 8 --ep_size 1 --draft_length 3 --output_length 4096 --engine TRTLLM --concurrency 32 --show_progress
```

Comment (benchislett) [reply examples/specdec_bench/README.md]: nevermind, I see now that this is the artifact directory for prepare_data.py

File: examples/specdec_bench/specdec_bench/datasets/speed.py [hunk 2]
```diff
@@ -0,0 +1,802 @@
...
+
+    @staticmethod
+    def _generate_textsort_prompt(prompt: str) -> str:
+        original_instruction = "\n    You are an AI assistant. Your job is to sort multiple book sections into the correct order.\n    Each time, you will be provided with 4 pieces of text.\n    These texts form a continuous part of a book, but are provided in random order.\n    You need to find the correct order and return the answer in a string.\n    For example, if you output [4, 1, 3, 2], that means the correct order is: Part 4 -> Part 1 -> Part 3 -> Part 2.\n    You will also be provided with the neighboring paragraphs before and after the 4 pieces of texts. \n\n    The case sample is shown below and you should give me the answer in the format exactly the same as the sample. \n\n    However, you should NOT focus on the content of sample answer. \n\n    Please do NOT output any extra content. \n    Sample Input (format only): \n\n    Before: XXX (Text before the continuous book part)\n\n\n    Part 1: XXX\n\n\n    Part 2: XXX\n\n\n    Part 3: XXX\n\n\n    Part 4: XXX\n\n\n    After: XXX (Text after the continuous book part)\n\n\n    Sample Output (format only): \n\n    Answer: [4, 1, 3, 2] \n\n\n\n"
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/datasets/speed.py:257]: Is this copied from some 3rd party GH repo?

File: examples/specdec_bench/specdec_bench/models/specbench_medusa.py
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/models/specbench_medusa.py:1]: License needs to be excluded from auto-header in pre-commit-config and written such that we have reference link (with commit hash) of file this is adapted from. Followed by original license / copyright and AFTER that, NVIDIA copyright/license

File: examples/specdec_bench/requirements_speed.txt
Comment (kevalmorabia97) [inline examples/specdec_bench/requirements_speed.txt:1]: Any reason to pin specific version? We want to avoid seeing vulnerability reported for older version in nspect scans so not having this pinned helps a lot. Unless there is a specific reason to keep this

File: .pre-commit-config.yaml
Comment (kevalmorabia97) [inline .pre-commit-config.yaml:1]: Why this file is excluded from lint/formatting checks?

File: examples/specdec_bench/specdec_bench/models/specbench_medusa.py
Comment (kevalmorabia97) [reply examples/specdec_bench/specdec_bench/models/specbench_medusa.py:1]: The file which has code copied from another 3rd party GitHub repository should first have a reference link (with commit hash) from which we copied the code, followed by original repo’s Copyright / License and finally ModelOpt Apache 2 Copyright / License header. You also need to exclude this file from license pre-commit hook so it doesnt auto-add ModelOpt license on the top of the file. See list of excluded files [here](https://github.com/NVIDIA/Model-Optimizer/blob/main/.pre-commit-config.yaml#L70-L101) and an example license header [here](https://github.com/NVIDIA/Model-Optimizer/blob/main/modelopt/torch/quantization/plugins/attention.py).

File: examples/specdec_bench/specdec_bench/metrics/specbench.py [hunk 3]
```diff
@@ -0,0 +1,237 @@
...
+
+from .acceptance_rate import AcceptanceRate
+
+
+class SpecBench(AcceptanceRate):
+    def __init__(self, requests):
+        super().__init__()
+        if not_installed:
+            raise ImportError(
+                "Please install rich, matplotlib, seaborn, and pandas to use the SpecBench metric"
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/metrics/specbench.py:41]: Can we have all these additional deps (not part of modelopt installation deps) mentioned in examples/specdec_bench/requirements.txt and mention in readme to install these deps before running the scripts?

File: examples/specdec_bench/specdec_bench/models/specbench_medusa.py
Comment (kevalmorabia97) [reply examples/specdec_bench/specdec_bench/models/specbench_medusa.py:1]: License header still not in correct order and missing link to original code file we copied from (including commit hash)

File: examples/specdec_bench/specdec_bench/models/specbench_medusa.py [hunk 4]
```diff
@@ -0,0 +1,284 @@
...
+import time
+
+import torch
+
+from .base import Model
+
+# Medusa dependencies from Spec-Bench
+try:
+    import os
+    import sys
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/models/specbench_medusa.py]: these can be moved out of the try block

File: examples/specdec_bench/specdec_bench/metrics/specbench.py (hunk 3 above)
Comment (IzzyPutterman) [reply examples/specdec_bench/specdec_bench/metrics/specbench.py:41]: Yep, already at the top of the readme for Speed bench running!

File: examples/specdec_bench/specdec_bench/utils.py [hunk 5]
```diff
@@ -19,12 +19,16 @@
 
 
 def get_tokenizer(path):
-    return AutoTokenizer.from_pretrained(path)
+    return AutoTokenizer.from_pretrained(path, trust_remote_code=True)
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/utils.py]: We cannot hard code `trust_remote_code=True` for security concerns. Please make this and argument to the function and require user to explicitly set it if required

File: examples/specdec_bench/README.md (hunk 1 above)
Comment (kevalmorabia97) [inline examples/specdec_bench/README.md]: Can you split into multiple lines per argument for better readability? 

Same for other commands in this file

File: examples/specdec_bench/requirements_speed.txt
Comment (kevalmorabia97) [reply examples/specdec_bench/requirements_speed.txt:1]: rich is already part of modelopt dependencies so not required here. For rest of deps, please avoid pinning upper version if not needed. I think `dataset<5` makes sense and is very less frequent major version release so we can keep it.

For seaborn and tiktoken, unless we use some special functions that may break, lets not pin upper version for them

File: examples/specdec_bench/specdec_bench/datasets/speed.py [hunk 6]
```diff
@@ -0,0 +1,807 @@
...
+from pathlib import Path
+from typing import Any, Literal, get_args
+
+from .base import Dataset, Request
+
+try:
+    import numpy as np
+    import pandas as pd
+    import tiktoken
+    from datasets import concatenate_datasets, load_dataset
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/datasets/speed.py:29]: We can add them to requirements.txt and remove the import guard here

File: examples/specdec_bench/specdec_bench/models/sglang.py [hunk 7]
```diff
@@ -43,43 +48,52 @@ def __init__(
             self.model = sgl.Engine(
                 model_path=model_dir,
                 skip_tokenizer_init=True,
-                mem_fraction_static=0.7,
-                disable_overlap_schedule=kwargs.get("disable_overlap_schedule", True),
+                trust_remote_code=True,
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/models/sglang.py]: same comment for all places we hardcode `trust_remote_code=True`

File: examples/specdec_bench/specdec_bench/datasets/speed.py (hunk 6 above)
Comment (IzzyPutterman) [reply examples/specdec_bench/specdec_bench/datasets/speed.py:29]: The goal is to not have to install any requirements to do standard runs if you dont want to.  Speed imports are hidden as to not force any installs

File: .pre-commit-config.yaml
Comment (IzzyPutterman) [reply .pre-commit-config.yaml:1]: Excluded because ruff is very unhappy with some of the white space in the system prompts and the name collision with datasets python package and datasets being a folder in the repo (even tho they dont collide)

File: examples/specdec_bench/specdec_bench/datasets/speed.py (hunk 6 above)
Comment (kevalmorabia97) [reply examples/specdec_bench/specdec_bench/datasets/speed.py:29]: How about we name these `requirements-speed.txt` and make it optional step in readme?

numpy is a default modelopt dependency and doesnt need to be kept in the try block

File: examples/specdec_bench/specdec_bench/metrics/specbench.py [hunk 8]
```diff
@@ -0,0 +1,237 @@
...
+import json
+import os
+from collections import defaultdict
+from itertools import chain
+from pathlib import Path
+from statistics import mean
+
+try:
+    import matplotlib.pyplot as plt
+    import pandas as pd
```
Comment (kevalmorabia97) [inline examples/specdec_bench/specdec_bench/metrics/specbench.py:25]: same for these 2 in `requirements-speed.txt`. Note that `rich` is default modelopt dependency so can be moved out of the try block

Comment (IzzyPutterman) [reply examples/specdec_bench/specdec_bench/metrics/specbench.py:25]: This isnt run in the modelopt env, so it needs to stay there.

File: examples/specdec_bench/specdec_bench/metrics/specbench.py (hunk 3 above)
Comment (kevalmorabia97) [reply examples/specdec_bench/specdec_bench/metrics/specbench.py:41]: In that case, lets not list individual dependencies in the error here instead let user know to install requirements file

File: examples/specdec_bench/specdec_bench/datasets/speed.py (hunk 6 above)
Comment (IzzyPutterman) [reply examples/specdec_bench/specdec_bench/datasets/speed.py:29]: [same as below] This isnt run in the modelopt env, so it needs to stay there.

File: examples/specdec_bench/requirements_speed.txt
Comment (IzzyPutterman) [reply examples/specdec_bench/requirements_speed.txt:1]: Updated
